<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-Malm24/blob/main/Semana_3_Actividad_1_A01113135.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Semana 3 - Actividad 1
Marco Antonio López Millán A01113135
Ciencia y analítica de datos (Gpo 10)
Jobish Vallikavungal Devassia


# **Fundamentos de bases de datos y para ciencia de datos.**

 La información puede ser recopilada en distintas formas y estructura:

Datos no estructurados: Documentos de texto, imagenes y audios.
Datos estructurada: datos que se encuentran organizados en formato de tabla.

Actualmente SQL domina la industria de los datos estructurados y es el sistema de bases de datos que normalmente se encuentra en las empresas.

Una base de datos es una tabla con indentificador de filas y encabezados para sus columnas. Estas tablas almacenan distintos tipo de datos y en diferente niveles de detalle.

A las filas de una base de datos regularmente se les llama record. Mientras que a las columnas se les puede llamar atributos o campos.

Primary Key es el nombre asignado a un valor unico que permite indentificar una fila.

Foreign key es cuando utilizamos una primary key  como referencia en una nueva tabla.

Normalización de base de datos: es el concepto utilizado para evitar el registro de datos redundantes y así reducir la cantidad de datos almacenados.

# **Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.**

Un Data Warehouse puede ser descrito como un base de datos gigante que integra información de distintas fuentes.  

El mayor reto de un data warehouse es tomar la información de distintas fuentes y habilitar que las personas que consumen dicha información puedan consultar diversas fuentes de manera simultánea. Para lograr esto es necesario lidiar con información faltante, mantener las escalas, terminología y otros temas importantes.

ETL consiste en transformar un grupo de data de manera offline para posteriormente importarla y cargarla en un data warehouse. Por su parte ELT consiste en extraer información cruda, cargarla en un sistema data warehouse y utilizar el poder del mismo para transformar la información.



In [27]:
import pandas as pd #Importando panda
import numpy as np  #Importando Numpy

In [28]:
df =pd.read_csv("/default of credit card clients.csv") #creando un DF

In [29]:
df #mostrando el contenido de DF

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [30]:
nueva_df=df.copy() #Realizamos una copia de DF
nueva_df

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [34]:
nueva_df.rename(columns = {"X1":"Monto Credito","X2":"Gen","X3":"Edu","X4":"E.Civil","X5":"Edad","X6":"H.sep","X7":"H.ago","X8":"H.jul","X9":"H.jun","X10":"H.may","X11":"H.abr","X12":"AB.sep","X13":"AB.ago","X14":"AB.jul","X15":"AB.jun","X16":"AB.may","X17":"AB.abr","X18":"AP.sep","X19":"AP.ago","X20":"AB.jul","X21":"AB.jun","X22":"AB.may","X23":"AB.abr","Y":"Resp"}, inplace = True)
#En este código nos encontramos reemplazando la variables por el un nombre facil de indentificar

In [35]:
nueva_df.isnull().sum().sort_values(ascending=True) #Nos permite identificar los valores nulos para cada variable y ordenarlos de menor a mayor

ID                0
Monto Credito     0
Gen               1
Edu               2
E.Civil           2
H.sep             3
Resp              3
H.ago             5
AB.abr            5
Edad              5
H.jul             7
AB.jul            8
AP.sep            8
H.jun             9
AP.ago            9
AB.abr           10
AB.may           11
AB.jun           11
AB.sep           11
AB.ago           11
AB.jul           13
H.abr            14
AB.jun           15
H.may            16
AB.may           17
dtype: int64

In [37]:
totalnulos = nueva_df.isnull().sum().sum() #Nos permite calcular el total de valores nulos
totalnulos 

196

In [51]:
totalfilas = len(nueva_df.index) #nos permite 
totalfilas

30000

In [52]:
print ("La proporción de datos nulos contra filas existentes es:", totalnulos/totalfilas*100, "%")

La proporción de datos nulos contra filas existentes es: 0.6533333333333333 %


Debido a que el máximo de columnas a eliminar asumiendo que los 196 nulos se encuentran en filas distinta representa un 0.65% de los datos, la estrategia más sencilla de aplicar es eliminar la filas donde encontramos nulos y esto tendría un impacto muy insignificante en el análisis de los datos.

In [53]:
nueva_df.dropna(axis = 0, inplace = True) #Eliminamos aquellas filas que cuentan con valores nulos
nueva_df

ID  Monto Credito  Gen  Edu  E.Civil  Edad  H.sep  H.ago  H.jul  \
0          1          20000  2.0  2.0      1.0  24.0    2.0    2.0   -1.0   
1          2         120000  2.0  2.0      2.0  26.0   -1.0    2.0    0.0   
2          3          90000  2.0  2.0      2.0  34.0    0.0    0.0    0.0   
3          4          50000  2.0  2.0      1.0  37.0    0.0    0.0    0.0   
4          5          50000  1.0  2.0      1.0  57.0   -1.0    0.0   -1.0   
...      ...            ...  ...  ...      ...   ...    ...    ...    ...   
29995  29996         220000  1.0  3.0      1.0  39.0    0.0    0.0    0.0   
29996  29997         150000  1.0  3.0      2.0  43.0   -1.0   -1.0   -1.0   
29997  29998          30000  1.0  2.0      2.0  37.0    4.0    3.0    2.0   
29998  29999          80000  1.0  3.0      1.0  41.0    1.0   -1.0    0.0   
29999  30000          50000  1.0  2.0      1.0  46.0    0.0    0.0    0.0   

       H.jun  ...   AB.jun   AB.may   AB.abr   AP.sep   AP.ago   AB.jul  \
0       -1.0  ...      0.0      0.0      0.0      0.0    689.0      0.0   
1        0.0  ...   3272.0   3455.0   3261.0      0.0   1000.0   1000.0   
2        0.0  ...  14331.0  14948.0  15549.0   1518.0   1500.0   1000.0   
3        0.0  ...  28314.0  28959.0  29547.0   2000.0   2019.0   1200.0   
4        0.0  ...  20940.0  19146.0  19131.0   2000.0  36681.0  10000.0   
...      ...  ...      ...      ...      ...      ...      ...      ...   
29995    0.0  ...  88004.0  31237.0  15980.0   8500.0  20000.0   5003.0   
29996   -1.0  ...   8979.0   5190.0      0.0   1837.0   3526.0   8998.0   
29997   -1.0  ...  20878.0  20582.0  19357.0      0.0      0.0  22000.0   
29998    0.0  ...  52774.0  11855.0  48944.0  85900.0   3409.0   1178.0   
29999    0.0  ...  36535.0  32428.0  15313.0   2078.0   1800.0   1430.0   

       AB.jun   AB.may  AB.abr  Resp  
0         0.0      0.0     0.0   1.0  
1      1000.0      0.0  2000.0   1.0  
2      1000.0   1000.0  5000.0   0.0  
3      1100.0   1069.0  1000.0   0.0  
4      9000.0    689.0   679.0   0.0  
...       ...      ...     ...   ...  
29995  3047.0   5000.0  1000.0   0.0  
29996   129.0      0.0     0.0   0.0  
29997  4200.0   2000.0  3100.0   1.0  
29998  1926.0  52964.0  1804.0   1.0  
29999  1000.0   1000.0  1000.0   1.0  

[29958 rows x 25 columns]

**¿Qué datos considero mas importantes?**
Amount of the given credit
History of past payment
Amount of bill statement
Amount of previous payment
  
**¿Por qué?**

Creo que el compartamiento de los últimos meses puede permitirnos hacer un pronóstico del comportamiento esperado del cliente. Sin embargo, variables como el género, educación, estado marital y la edad son variables que regularmente son tomados en cuenta para el análisis crediticio de una persona ya que permiten agruparlos e indentificar si existen comportamientos predecibles de acuerdo a su caracteristicas. 

Ejemplo: podría ser que regularmente los hombre con una edad mayor 35 que se encuentran solteros tienden a ser clientes morosos.
 
Por lo tanto considero que es importante no perder de vista este tipo de información sobre los clientes.


**¿Se eliminaron o reemplazaron datos nulos?**

En esta ocasión se decidió eliminar todas la filas donde se encontraban datos nulos, por lo tanto debido a la limpieza se pasó de contar con 30000 registros 29958. 

Otra solución pudo ser el sustituir los valores faltantes con la media, mediana o moda se los registros que se encontraban completos. 
 
Sin embargo se optó por la elimincación por lista.

**¿Qué se hizo y por qué?**

Se eliminaron debido al análisis que se realizó donde existian 196 datos nulos, suponiendo que cada uno de ellos se hubiera encontrado en una fila distinta los datos hubieran representado un 0.65% lo cual es insignificante.

En realidas tras realizar la limpieza solo se eliminaron 42 filas lo cual representa 0.0014%.

Lo cual nos indica que la muestra es suficientemente grande como para evitar que las estimaciones arrojados por el modelo puedan tener algún sesgo y esto nos permite llegar a resultados conservadores.

**¿Es necesario limpiar los datos para el análisis? **
Sí, es necesario.
**¿Por qué?**

El contar con datos nulos puedes sesgar nuestros cálculos o modelos.

Además es importante mencionar que la mayoría de los algoritmos utilizados para el análisis de datos no funcionan cuando existe valores nulos o faltantes, por lo tomó mayor relevancia el encontrar una forma de manejar dicha informacións. Los datos faltantes pueden producir resultados sesgados, lo que lleva a conclusiones no válidas.


**¿Existen problemas de formato que deban solucionar antes del proceso de modelado? Sí / No / Por qué.**

Una de las cosas que podría ser considerada un problema para un manejo sencillo de la información fue el título de las columnas. Al contar con 23 variables  el hecho de no tener un nombre con el cual pudiéramos indentificar sencillamente  el dato observado podría complicar el análisis. Por ello se decidió dar un nombre relacionado con su descripción.

**¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?**

En este ejercicio para el proceso de limpieza se decidió optar por la eliminación por lista que consiste en eliminar aquellos registros con datos faltantes y además se cambió el título de los atributos (columnas) para facilitar su manejo.
